### PR TITLE
Redux saga

### PR DESCRIPTION
**Pull request by dfcz3322**

### CODE REVIEW

File: src/actions/comics.ts [hunk 1]
```diff
@@ -1,9 +1,34 @@
 import { IMarvelEntityResponse } from '../interface/interface';
+
+
 export const SET_COMICS = 'COMICS/SET_COMICS';
```
Comment (alinalaktionova) [inline src/actions/comics.ts]: again without COMICS/

File: src/actions/comics.ts [hunk 2]
```diff
@@ -1,9 +1,34 @@
...
+export const GET_COMICS = 'GET_COMICS';
+export const SET_COMICS_ERROR = 'SET_COMICS_ERROR';
```
Comment (alinalaktionova) [inline src/actions/comics.ts]: why all types start from SET?)) Usually we have 3 types of action getComics, getComicsSuccess and getComicsError

File: src/actions/comics.ts [hunk 3]
```diff
@@ -1,9 +1,34 @@
...
+export const SET_IS_LOADING = 'SET_IS_LOADING';
 
 export const setComics = (comics: IMarvelEntityResponse[]): { type: string; payload: IMarvelEntityResponse[] } => {
   return {
     type: SET_COMICS,
     payload: comics,
   };
 };
+
+export const getComics = (id: string): { type: string; payload: string } => {
```
Comment (alinalaktionova) [inline src/actions/comics.ts]: you can create separate type for a return value

File: src/actions/comics.ts [hunk 4]
```diff
@@ -1,9 +1,34 @@
...
+
+export const setComicsError = (): { type: string } => {
+  return {
+    type: SET_COMICS_ERROR,
+  };
+};
+
+export const setIsLoading = (isLoading: boolean): { type: string; payload: boolean } => {
+    return {
+        type: SET_IS_LOADING,
```
Comment (alinalaktionova) [inline src/actions/comics.ts]: tabs

File: src/actions/search.ts [hunk 5]
```diff
@@ -3,11 +3,53 @@ interface ISetCharacters {
...
+  type: string;
+  payload: string;
+}
+
+interface IGetHeroes {
+  type: string;
+  payload: string;
+}
+
 export const SET_CHARACTERS = 'SET_CHARACTERS';
```
Comment (alinalaktionova) [inline src/actions/search.ts]: characters or heroes? choose one

File: src/actions/search.ts [hunk 6]
```diff
@@ -3,11 +3,53 @@ interface ISetCharacters {
...
+export const GET_HEROES = 'GET_HEROES';
+export const SET_HEROES_ERROR = 'SET_HEROES_ERROR';
```
Comment (alinalaktionova) [inline src/actions/search.ts]: the same with setters

File: src/actions/search.ts [hunk 7]
```diff
@@ -3,11 +3,53 @@ interface ISetCharacters {
...
+export const SET_SEARCH_QUERY = 'SET_SEARCH_QUERY';
```
Comment (alinalaktionova) [inline src/actions/search.ts]: why search query is in redux again?)

File: src/reducer/comicsReducer.tsx [hunk 8]
```diff
@@ -1,21 +1,29 @@
...
+  action: { type: string; payload: unknown }
 ): IComicsReduxState => {
   switch (action.type) {
     case SET_COMICS:
-      return { ...state, comics: action.payload };
+      return { ...state, comics: action.payload as IMarvelEntityResponse[], hasError: false };
+    case SET_COMICS_ERROR:
+      return { ...state, comics: [], hasError: true };
+      case SET_IS_LOADING:
+      return {...state, isLoading: action.payload as boolean };
```
Comment (alinalaktionova) [inline src/reducer/comicsReducer.tsx]: better to set true to loading state right here

File: src/sagas.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+interface TaskAction extends Action, ITask {
+  type: 'TASK_ADD';
+}
+
+function* fetchHeroes(action: { payload: string }): any {
+  try {
+    yield put(setIsHeroesLoading(true));
+    const heroes = yield call(<any>handleRequest, `characters`, action.payload);
```
Comment (alinalaktionova) [inline src/sagas.ts]: any

File: src/sagas.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+import { Action } from 'redux';
+import { call, put, takeLatest } from 'redux-saga/effects';
+import { handleRequest } from './api/requestHandler';
+import { GET_COMICS, setComics, setComicsError, setIsLoading as setIsComicsLoading  } from './actions/comics';
+
+interface ITask {
+  payload: string;
+}
+
+interface TaskAction extends Action, ITask {
```
Comment (alinalaktionova) [inline src/sagas.ts:12]: for what?

File: src/sagas.ts [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    const heroes = yield call(<any>handleRequest, `characters`, action.payload);
+    yield put(setCharacters(heroes));
+  } catch (e) {
+    yield put(setHeroesError());
+  } finally {
+    yield put(setIsHeroesLoading(false))
+  }
+}
+
+function* fetchComics(action: { payload: string }): any {
```
Comment (alinalaktionova) [inline src/sagas.ts]: remove any

**Review by alinalaktionova — CHANGES_REQUESTED** [review]

File: src/sagas.ts (hunk 9 above)
Comment (dfcz3322) [reply src/sagas.ts]: there are issues with saga's typings and I'm not sure how to resolve them. Could you please give some advice?

File: src/sagas.ts (hunk 10 above)
Comment (dfcz3322) [reply src/sagas.ts:12]: It's a workaround for typing issues. `takeLatest` throws errors without it

File: src/sagas.ts (hunk 11 above)
Comment (dfcz3322) [reply src/sagas.ts]: Same issues, generator functions are really hard to type properly. Please, help me!

**Review by alinalaktionova — APPROVED** [review]

**Pull request merged**